File: esercitazioni/java/proxySkeleton/magazzino/src/coda/impl/CodaCircolare.java
package coda.impl;

import coda.Coda;

public class CodaCircolare implements Coda {
    private final int[] buffer;
    private int testa;
    private int coda;
    private int size;

    public CodaCircolare(int capacity) {
        this.buffer = new int[capacity];
        this.testa = 0;
        this.coda = 0;
        this.size = 0;
    }

    public int preleva() {
        size++;
        return buffer[testa++ % buffer.length];
    }

    public void deposita(int value) {
        size--;
        buffer[coda++ % buffer.length] = value;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isFull() {
        return size == buffer.length;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public int getCapacity() {
        return buffer.length;
    }

}

File: esami/2024-06-26_grpc+flask_REBOOT_java/manager_python/queues/codaCircolare.py
from queues.queue import IQueue


class CodaCircolare(IQueue):
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.buffer = [None] * capacity
        self.testa = 0
        self.coda = 0
        self.size = 0

    def push(self, item):
        self.buffer[self.testa] = item
        self.testa = (self.testa + 1) % self.capacity
        self.size += 1

    def pop(self):
        item = self.buffer[self.coda]
        self.coda = (self.coda + 1) % self.capacity
        self.size -= 1
        return item

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size == self.capacity

    def __len__(self) -> int:
        return self.size
